Support for ED25519 ssh keys?

When using ssh with an ed25519 key, the app crashes and I see the following stacktrace in logcat:

`12-02 21:26:06.744 31014 31040 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
12-02 21:26:06.744 31014 31040 E AndroidRuntime: Process: de.eidottermihi.raspicheck, PID: 31014
12-02 21:26:06.744 31014 31040 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:325)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:761)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: Caused by: java.lang.UnsupportedOperationException: Don't know how to decode key:ssh-ed25519
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.common.KeyType$4.readPubKeyFromBuffer(KeyType.java:191)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.common.Buffer.readPublicKey(Buffer.java:430)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.userauth.keyprovider.OpenSSHKeyFile.initPubKey(OpenSSHKeyFile.java:98)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.userauth.keyprovider.OpenSSHKeyFile.init(OpenSSHKeyFile.java:68)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.userauth.keyprovider.PKCS8KeyFile.init(PKCS8KeyFile.java:107)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.SSHClient.loadKeys(SSHClient.java:493)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at net.schmizz.sshj.SSHClient.loadKeys(SSHClient.java:448)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at de.eidottermihi.rpicheck.ssh.impl.RaspiQuery.connectWithPubKeyAuth(RaspiQuery.java:1106)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at de.eidottermihi.rpicheck.activity.SSHQueryTask.doInBackground(SSHQueryTask.java:89)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at de.eidottermihi.rpicheck.activity.SSHQueryTask.doInBackground(SSHQueryTask.java:46)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:305)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
12-02 21:26:06.744 31014 31040 E AndroidRuntime: 	... 4 more`

Is there a way to support these keys? If not, it seems like the app should alert the user rather than blindly crash.

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -37,8 +37,8 @@ android {
     buildTypes {
         release {
             shrinkResources true
-            minifyEnabled true
-            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-project.txt'
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt')//, 'proguard-project.txt'
         }
     }
 

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -28,8 +28,9 @@ android {
 
     defaultConfig {
         applicationId "de.eidottermihi.raspicheck"
-        minSdkVersion 9
+        minSdkVersion 21
         targetSdkVersion 23
+        multiDexEnabled true
         versionName version
     }
 

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -37,8 +37,14 @@ android {
     buildTypes {
         release {
             shrinkResources true
-            minifyEnabled false
-            proguardFiles getDefaultProguardFile('proguard-android.txt')//, 'proguard-project.txt'
+            minifyEnabled true
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-project.txt'
+        }
+
+        debug {
+            shrinkResources false
+            minifyEnabled true
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-project.txt'
         }
     }
 

--- a/app/proguard-project.txt
+++ b/app/proguard-project.txt
@@ -22,30 +22,50 @@
 # Dont obfuscate
 -dontobfuscate
 
+-dontwarn com.google.errorprone.annotations.**
 # Libraries
 # SSHJ
 -keep class net.schmizz.** { *; }
 -keep interface net.schmizz.** { *; }
 -keep class com.jcraft.** { *; }
 -keep interface com.jcraft.** { *; }
+-keep class com.hierynomus.sshj.** { *; }
+-keep interface com.hierynomus.sshj.** { *; }
+-keep class djb.** { *; }
+-keep interface djb.** { *; }
+-keep class net.i2p.** { *; }
+-keep interface net.i2p.** { *; }
 -keep class org.spongycastle.** { *; }
 -keep interface org.spongycastle.** { *; }
+-keep class org.bouncycastle.** { *; }
+-keep interface org.bouncycastle.** { *; }
+-dontwarn org.ietf.**
 # were not using ldap
 -dontwarn org.spongycastle.jce.provider.X509LDAPCertStoreSpi
 -dontwarn org.spongycastle.x509.util.LDAPStoreHelper
+# no GssApi Auth
+-dontwarn javax.security.auth.**
+# no JNDI
+-dontwarn javax.naming.**
 # PrettyTime
 -keep class org.ocpsoft.prettytime.** { *; }
 -keep interface org.ocpsoft.prettytime.** { *; }
 # Logback
 -keep class ch.qos.logback.** { *; }
 -keep interface ch.qos.logback.** { *; }
+-dontwarn java.nio.file.**
 # we're not using remote logging
--dontwarn ch.qos.logback.core.net.*
+-dontwarn ch.qos.logback.core.net.**
 # slf4j
 -keep class org.slf4j.** { *; }
 -keep interface org.slf4j.** { *; }
 # guava
+-dontwarn com.google.j2objc.annotations.**
+-dontwarn com.google.appengine.**
+-dontwarn com.google.apphosting.**
+-dontwarn java.lang.ClassValue
 -dontwarn sun.misc.Unsafe
+-dontwarn org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement
 
 # Submodule Libraries
 -keep class de.eidottermihi.rpicheck.ssh.** { *; }

--- a/ssh/build.gradle
+++ b/ssh/build.gradle
@@ -10,10 +10,15 @@ targetCompatibility = JavaVersion.VERSION_1_7
 
 
 dependencies {
-    compile 'com.google.guava:guava:18.0'
+    compile 'com.google.guava:guava:20.0'
     compile 'org.ocpsoft.prettytime:prettytime:2.1.2.Final'
     compile 'org.slf4j:slf4j-api:1.7.7'
-    compile 'de.eidottermihi:sshj4android:0.11.0'
+    compile 'com.hierynomus:sshj:0.22.0'
+    compile 'com.madgag.spongycastle:core:1.56.0.0'
+    compile 'com.madgag.spongycastle:prov:1.56.0.0'
+    compile 'com.madgag.spongycastle:bcpkix-jdk15on:1.56.0.0'
+    compile 'com.madgag.spongycastle:bcpg-jdk15on:1.56.0.0'
+
 
     testCompile 'junit:junit:4.12'
     testCompile 'org.mockito:mockito-all:1.9.5'

--- a/ssh/build.gradle
+++ b/ssh/build.gradle
@@ -13,7 +13,7 @@ dependencies {
     compile 'com.google.guava:guava:20.0'
     compile 'org.ocpsoft.prettytime:prettytime:2.1.2.Final'
     compile 'org.slf4j:slf4j-api:1.7.7'
-    compile 'com.hierynomus:sshj:0.22.0'
+    compile 'com.hierynomus:sshj:0.23.0'
     compile 'com.madgag.spongycastle:core:1.56.0.0'
     compile 'com.madgag.spongycastle:prov:1.56.0.0'
     compile 'com.madgag.spongycastle:bcpkix-jdk15on:1.56.0.0'

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/IQueryService.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/IQueryService.java
@@ -182,8 +182,6 @@ public interface IQueryService {
      */
     public abstract void disconnect() throws RaspiQueryException;
 
-    public abstract String getFile(String path);
-
     public abstract String getHostname();
 
     public abstract void setHostname(String hostname);

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
@@ -31,10 +31,12 @@ import net.schmizz.sshj.transport.TransportException;
 import net.schmizz.sshj.userauth.UserAuthException;
 import net.schmizz.sshj.userauth.keyprovider.KeyProvider;
 
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.security.Security;
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
@@ -102,6 +104,11 @@ public class RaspiQuery implements IQueryService {
     private String hostname;
     private String username;
     private int port = DEFAULT_SSH_PORT;
+
+    static {
+        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+        Security.addProvider(new org.spongycastle.jce.provider.BouncyCastleProvider());
+    }
 
     /**
      * Initialize a new RaspiQuery.

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
@@ -36,10 +36,12 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.security.Provider;
 import java.security.Security;
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -118,6 +120,18 @@ public class RaspiQuery implements IQueryService {
      * @param port ssh port to use (if null, default will be used)
      */
     public RaspiQuery(final String host, final String user, final Integer port) {
+        final Provider[] providers = Security.getProviders();
+        LOGGER.debug("Registered JCE providers...");
+        for (Provider prov : providers
+                ) {
+            LOGGER.debug("Provider: {} - {}", prov.getName(), prov.getInfo());
+        }
+        final Set<String> signatures = Security.getAlgorithms("signature");
+        LOGGER.debug("Availabe signatures...");
+        for (String sig : signatures
+                ) {
+            LOGGER.debug("Signature: {}", sig);
+        }
         if (Strings.isNullOrEmpty(host)) {
             throw new IllegalArgumentException("hostname should not be blank.");
         } else if (Strings.isNullOrEmpty(user)) {

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/RaspiQuery.java
@@ -22,7 +22,6 @@ import com.google.common.base.Optional;
 import com.google.common.base.Splitter;
 import com.google.common.base.Strings;
 
-import net.schmizz.sshj.AndroidConfig;
 import net.schmizz.sshj.SSHClient;
 import net.schmizz.sshj.common.IOUtils;
 import net.schmizz.sshj.connection.ConnectionException;
@@ -31,14 +30,11 @@ import net.schmizz.sshj.connection.channel.direct.Session.Command;
 import net.schmizz.sshj.transport.TransportException;
 import net.schmizz.sshj.userauth.UserAuthException;
 import net.schmizz.sshj.userauth.keyprovider.KeyProvider;
-import net.schmizz.sshj.xfer.FileSystemFile;
-import net.schmizz.sshj.xfer.scp.SCPFileTransfer;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
-import java.security.Security;
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
@@ -106,11 +102,6 @@ public class RaspiQuery implements IQueryService {
     private String hostname;
     private String username;
     private int port = DEFAULT_SSH_PORT;
-
-    static {
-        //Security.removeProvider("BC");
-        Security.insertProviderAt(new org.spongycastle.jce.provider.BouncyCastleProvider(), 1);
-    }
 
     /**
      * Initialize a new RaspiQuery.
@@ -1187,28 +1178,6 @@ public class RaspiQuery implements IQueryService {
     /*
      * (non-Javadoc)
      *
-     * @see de.eidottermihi.rpicheck.ssh.IQueryService#getFile(java.lang.String)
-     */
-    @Override
-    public final String getFile(String path) {
-        if (client != null) {
-            if (client.isConnected()) {
-                SCPFileTransfer scpClient = new SCPFileTransfer(client);
-                try {
-                    scpClient.download("/boot/config.txt", new FileSystemFile(
-                            "/"));
-                } catch (IOException e) {
-                    e.printStackTrace();
-                }
-
-            }
-        }
-        return null;
-    }
-
-    /*
-     * (non-Javadoc)
-     *
      * @see de.eidottermihi.rpicheck.ssh.IQueryService#getHostname()
      */
     @Override
@@ -1274,7 +1243,7 @@ public class RaspiQuery implements IQueryService {
      */
 
     public SSHClient newAndroidSSHClient() {
-        return new SSHClient(new AndroidConfig());
+        return new SSHClient(new SpongyCastleAndroidConfig());
     }
 
     @Override

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
@@ -1,0 +1,19 @@
+package de.eidottermihi.rpicheck.ssh.impl;
+
+import net.schmizz.sshj.AndroidConfig;
+import net.schmizz.sshj.Config;
+import net.schmizz.sshj.common.SecurityUtils;
+import net.schmizz.sshj.transport.random.BouncyCastleRandom;
+import net.schmizz.sshj.transport.random.SingletonRandomFactory;
+
+/**
+ * Created by Michael on 28.08.2017.
+ */
+
+class SpongyCastleAndroidConfig extends AndroidConfig {
+    public SpongyCastleAndroidConfig() {
+        super();
+        initKeyExchangeFactories(true);
+        initFileKeyProviderFactories(true);
+    }
+}

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
@@ -1,16 +1,40 @@
 package de.eidottermihi.rpicheck.ssh.impl;
 
-import net.schmizz.sshj.AndroidConfig;
-import net.schmizz.sshj.Config;
+import com.hierynomus.sshj.signature.SignatureEdDSA;
+
+import net.schmizz.sshj.DefaultConfig;
 import net.schmizz.sshj.common.SecurityUtils;
-import net.schmizz.sshj.transport.random.BouncyCastleRandom;
+import net.schmizz.sshj.signature.SignatureDSA;
+import net.schmizz.sshj.signature.SignatureRSA;
+import net.schmizz.sshj.transport.random.JCERandom;
 import net.schmizz.sshj.transport.random.SingletonRandomFactory;
 
-public class SpongyCastleAndroidConfig extends AndroidConfig {
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+import java.security.Security;
+
+public class SpongyCastleAndroidConfig extends DefaultConfig {
+
+    static {
+        SecurityUtils.registerSecurityProvider("org.spongycastle.jce.provider.BouncyCastleProvider");
+    }
+
     public SpongyCastleAndroidConfig() {
         super();
         initKeyExchangeFactories(true);
         initRandomFactory(true);
         initFileKeyProviderFactories(true);
+    }
+
+    // don't add ECDSA
+    protected void initSignatureFactories() {
+        setSignatureFactories(new SignatureRSA.Factory(), new SignatureDSA.Factory(),
+                // but add EdDSA !
+                new SignatureEdDSA.Factory());
+    }
+
+    @Override
+    protected void initRandomFactory(boolean ignored) {
+        setRandomFactory(new SingletonRandomFactory(new JCERandom.Factory()));
     }
 }

--- a/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
+++ b/ssh/src/main/java/de/eidottermihi/rpicheck/ssh/impl/SpongyCastleAndroidConfig.java
@@ -6,14 +6,11 @@ import net.schmizz.sshj.common.SecurityUtils;
 import net.schmizz.sshj.transport.random.BouncyCastleRandom;
 import net.schmizz.sshj.transport.random.SingletonRandomFactory;
 
-/**
- * Created by Michael on 28.08.2017.
- */
-
-class SpongyCastleAndroidConfig extends AndroidConfig {
+public class SpongyCastleAndroidConfig extends AndroidConfig {
     public SpongyCastleAndroidConfig() {
         super();
         initKeyExchangeFactories(true);
+        initRandomFactory(true);
         initFileKeyProviderFactories(true);
     }
 }

